Implementation of DPL serializer using boost serialization routines

gabrielefronze opened the pull request:
Implementation of A DPL-compliant dataset serialiser.
It works flawlessly for STL containers such as vectors, arrays and lists.
It can be used for raw arrays.
This utility makes use of boost serialisation routines.
Various tests are performed on the DPL-independent methods.
A DPL workflow will be added to the tests to perform top-level checks.

sawenzel commented on Detectors/MUON/MID/Base/include/MIDBase/DPLSerializer.h:
```diff
@@ -0,0 +1,111 @@
+// or submit itself to any jurisdiction.
+
+/// \file   MIDBase/DPLSerializer.h
+/// \brief  Templated boost serializer/deserializer for MID vectors in DPL
```
Nice ideas.

A few comments related to this file (and maybe others):

a) most certainly it should go into the Framework directory

b) `DPLSerializer` sounds very generic and gives the impression that it could use a mechanism different than boost. Is this true? Otherwise, I suggest to call it `BoostDPLSerializer` or similar.

c) Is it possible to static_assert when some data structure did not implement a boost serializer?

d) Can you say something about the performance gain (difference) compared to ROOT serialization?

gabrielefronze replied:
Ciao Sandro,

Here my answers:

a) I wanted to put this DPL-serializer alongside the one Diego already pushed. Of course I can move it to the Framework folder, maybe in the Utils sub-one.

b) I see the point and indeed this helper uses boost serialisation. I will consider your proposal and change the naming following the advice.

c) Yes, sure. I'll do that.

d) I did not perform tests about that. I can manage to provide some measurements in the next days!

gabrielefronze replied:
Ciao Matthias,

Thanks for the comment.
I will work on b and c and update the branch asap (with some squashes as well).
I'll dig a bit on where to relocate this tool.
For the benchmark Vs ROOT serialisation I have to learn something about it, but I'll try to provide a similar implementation using that.

Just an additional update: I have tested the tool with a workflow and everything works as intended.

Cheers, and happy holidays!

mkrzewic replied:
Hi @matthiasrichter @gabrielefronze wouldn't it be nice if these things would not be tied too closely to DPL? A DPL API can always be cooked and made dependent on this, but there might be use cases outside this high level API...

gabrielefronze replied:
Hi,

The core serialising method is NOT dependent on DPL, but on boost.
Basically anyone can implement a wrapper of the wanted API that allocates the message space and serialises the container automagically using the core call.
In fact, as you can see, the SerialiseContainer method has no DPL calls within.
Maybe the SerialiseContainer implementation can be moved in DataFormats, so that every "framework" can implement the specialised call.
Let's see what are the others saying about that.
Cheers

gabrielefronze commented:
Dear @sawenzel @matthiasrichter @mkrzewic and all,

As you can see I have done some work on this PR.

- I have protected the boost serializer against wrong usage via`std::enable_if` and static asserts as suggested by @sawenzel (and it was a tricky task!). by doing so the class can basically work with any container handled by boost.
- The core BoostSerializer has been moved to CommonUtils bucket, since as suggested by @mkrzewic and @matthiasrichter  this is nor a DPL neither an MID tool, but a general one. Tests have been moved as well.
- I have added a DPLBoostSerializer wrapper in FrameworkCore, alongside TMessageSerializer.

I'll handle Travis errors tomorrow!

Cheers,

Gabriele

gabrielefronze commented:
Dear all,

As you can see I made Travis happy, but code readability (`static assert` and `enable_if` conditions) are much less readable in my opinion.
Feel free to pick (or not) the last commit: I won't squash it unless you ask for it.
Cheers!

Gabriele

dennisklein commented:
'Boost serialization' rings a bell, have you seen @NicolasWinckler's work in FairRoot: https://github.com/FairRootGroup/FairRoot/blob/dev/base/MQ/policies/Serialization/BoostSerializer.h? I have not looked at it in detail though.

gabrielefronze commented:
Ciao @dennisklein,

I saw this implementation, but it did not fit my specific requirements.
I have to admit that TClonesArray handling would be a nice addition, my goal with this PR is rather to introduce an unspecialised serializer which can be easily expanded and that can be wrapped in several message passing frameworks.
In fact the current implementation is currently used in the DPL through my implementation, but you can easily see that @NicolasWinckler's solution can become a FairMQDevice wrapper of my serializer, adding to it some interesting features (STL containers serialisation OOTB).

I think that a centralised serialisaton/deserialisation tool can make the difference in this situation by helping the user through type checks which are implemented once and in one place.

What do you think about that?

Cheers!

dennisklein commented:
> but you can easily see that @NicolasWinckler's solution can become a FairMQDevice wrapper of my serializer, adding to it some interesting features (STL containers serialisation OOTB).

I agree, that would be nice.

> I think that a centralised serialisaton/deserialisation tool can make the difference in this situation by helping the user through type checks which are implemented once and in one place.

I was also thinking in the direction, where to put your code (+ parts of Nicolas' code) - repo-wise. But I don't have a better idea.

> but code readability (`static assert` and `enable_if` conditions) is much worse.

You could declare `_v`/`_t` template aliases to shorten your meta function calls `::value`/`::type` as it is done in the STL since C++14 - not a huge improvement, but saves a few characters, e.g.:

Btw, why did Travis not like the line breaks?

What's the reason for using `boost::is_class` over `std::is_class`?

gabrielefronze commented:
Ciao @dennisklein,

Thanks for the discussion and the suggestions.
I have tried to revert the new lines so that we can figure out what's wrong with Travis. Indeed it seems to me the code is much better formatted in this version, hence the clang-format rules might be worth an update.

I have also replaced `boost::is_class` with `std::is_class`.

gabrielefronze commented:
Dear all,

I have implemented all the changes we discussed about and rebased on dev to solve the merge conflicts affecting CMake files.
wrt @sawenzel 4 points here you can find the outcome:
a) I think this has been addressed, since the original file has been split in a base serializer which has been moved to a common directory and the DPL wrapper added to the Framework dir.
b) The names have been updated. We could think about changing also the methods names. If you think so please tell.
c) This is done and almost every possible exception is now handled.
d) Not done yet.

I wish to point out that some lines in the `enable_if` conditions have been split for better code readability. After some digging (thanks to @dennisklein) we figured out that `clang-format` might be dumb about lines splitting, hence @ktf decided to keep it strict looking at the possible false positives 
[AliceO2Group/AliceO2@a149d16](https://github.com/AliceO2Group/AliceO2/commit/a149d1654a5827ff7c649515ff987f91a67e46c4).
This might be the case, so Travis test failure should be ignored.

Cheers

sawenzel commented:
@ktf: Can you take care of this PR as it is touching framework stuff?

ktf commented:
I will chat with Gabriele tomorrow

ktf commented:
This is fine as initial setup and we can merge it, but as discussed with @gabrielefronze we should:

* Implement something along the lines of `ROOTSerialized` (e.g. `BoostSerialized`) to force the serialization via boost.
* Add a `BoostObjectContext` along the lines of `ROOTObjectContext` to hold the intermediate buffers which have been serialised with boost and which have not been sent yet (e.g. to implement snapshot, which this version does not provide). This could possibly be a generic `RawBufferContext` which we could reused for other serialization methods as well.
* Follow the API of `TMessageSerializer` to comply with FairMQ serialization support.
* Implement a `BoostDataAllocator.h` where you keep the specialisations for `make`, `adopt` and `snapshot` of `DataAllocator`.

The end goal is to have the same API for Boost serialization as we have for ROOT or the standard buffers, i.e. something along the lines of:

```C++
std::unique_ptr<MIDClass> midObj = ctx.outputs().make<BoostSerialised<MIDClass>>();
```

which is then sent when the algorithm completes.
